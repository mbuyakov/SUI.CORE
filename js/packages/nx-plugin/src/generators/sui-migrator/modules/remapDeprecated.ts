import {factory, ImportDeclaration, QualifiedName} from "typescript";
import {logWithPrefix} from "../../../utils/logger";
import {astQuery, astReplace, parseImport, printNode} from "@sui/lib-typescript-ast";


export function remapDeprecated(content: string): string {
  let shouldAddImport = false;
  content = astReplace(content, "QualifiedName", (node: QualifiedName) => {
    if (node.left.getText() == "JSX") {
      shouldAddImport = true;
      logWithPrefix("remapDeprecated", `Replace JSX.${node.right.getText()} to React.JSX.${node.right.getText()}`);

      return printNode(
        factory.createQualifiedName(
          factory.createQualifiedName(
            factory.createIdentifier("React"),
            factory.createIdentifier("JSX")
          ),
          node.right
        )
      );
    }
  });

  // Check if import exist
  if (shouldAddImport) {
    let hasReactImport = false;
    astQuery<ImportDeclaration>(content, "ImportDeclaration").forEach(node => {
      const parsedImport = parseImport(node);
      if (parsedImport.from == "react" && (parsedImport.import == "React" || parsedImport.nsImport == "React")) {
        hasReactImport = true;
      }
    });

    shouldAddImport = !hasReactImport;
  }

  if (shouldAddImport) {
    let isFirst = true;
    content = astReplace(content, "ImportDeclaration", (node: ImportDeclaration) => {
      if (isFirst) {
        isFirst = false;
        return printNode(node) + "\nimport React from \"react\";";
      }
    });
  }

  return content;
}
