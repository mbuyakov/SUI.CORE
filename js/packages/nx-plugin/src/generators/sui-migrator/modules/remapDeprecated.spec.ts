import {remapDeprecated} from "./remapDeprecated";

describe("remap deprecated", () => {

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("should leave as is", () => {
    const consoleLog = jest.spyOn(console, "log");
    const source = `
import React from "react";
const a: React.JSX.Element = null;
    `.trim();
    expect(remapDeprecated(source)).toBe(source);
    expect(consoleLog).not.toBeCalled();
  });

  it("should replace JSX", () => {
    const consoleLog = jest.spyOn(console, "log");
    expect(remapDeprecated(`
import smth from "smth";
const a: JSX.Element = null;
    `.trim())).toMatchSnapshot();
    expect(consoleLog).toBeCalledTimes(1);
  });

  it("should leave import as is (default import)", () => {
    const consoleLog = jest.spyOn(console, "log");
    const source = `
import React from "react";
const a: JSX.Element = null;
    `.trim();
    expect(remapDeprecated(source)).toBe(source.replace("JSX.Element", "React.JSX.Element"));
    expect(consoleLog).toBeCalledTimes(1);
  });

  it("should leave import as is (namespace import)", () => {
    const consoleLog = jest.spyOn(console, "log");
    const source = `
import React from "react";
const a: JSX.Element = null;
    `.trim();
    expect(remapDeprecated(source)).toBe(source.replace("JSX.Element", "React.JSX.Element"));
    expect(consoleLog).toBeCalledTimes(1);
  });
});
