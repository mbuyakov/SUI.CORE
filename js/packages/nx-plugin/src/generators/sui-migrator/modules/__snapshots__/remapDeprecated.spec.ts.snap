// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`remap deprecated should replace JSX 1`] = `
"import smth from \\"smth\\";
import React from \\"react\\";
const a: React.JSX.Element = null;"
`;
